account-integrations/safe/src/SafeZkEmailRecoveryPlugin.sol

```diff
@@ -94,69 +126,146 @@ contract SafeZkEmailRecoveryPlugin {
      * @param c part of the proof
      */
-    function recoverAccount(
+    function initiateRecovery(
         address safe,
-        address ecdsaPlugin,
         address newOwner,
         string memory emailDomain,
         uint256[2] memory a,
         uint256[2][2] memory b,
         uint256[2] memory c
     ) external {
-        ZkEmailRecoveryStorage memory recoveryStorage = zkEmailRecoveryStorage[
-            safe
-        ];
+        RecoveryRequest memory recoveryRequest = recoveryRequests[safe];
+
+        if (recoveryRequest.executeAfter > 0) {
+            revert RECOVERY_ALREADY_INITIATED();
+        }
 
         if (
             !this.isDKIMPublicKeyHashValid(
                 safe,
                 emailDomain,
-                recoveryStorage.dkimPublicKeyHash
+                recoveryRequest.dkimPublicKeyHash
             )
         ) {
             revert INVALID_DKIM_KEY_HASH(
                 safe,
                 emailDomain,
-                recoveryStorage.dkimPublicKeyHash
+                recoveryRequest.dkimPublicKeyHash
             );
         }
 
         uint256[4] memory publicSignals = [
             uint256(uint160(safe)),
-            uint256(recoveryStorage.recoveryHash),
+            uint256(recoveryRequest.recoveryHash),
             uint256(uint160(newOwner)),
-            uint256(recoveryStorage.dkimPublicKeyHash)
+            uint256(recoveryRequest.dkimPublicKeyHash)
         ];
 
         // verify proof
         bool verified = verifier.verifyProof(a, b, c, publicSignals);
         if (!verified) revert INVALID_PROOF();
 
-        bytes memory data = abi.encodeWithSignature(
-            "enable(bytes)",
-            abi.encodePacked(newOwner)
-        );
+        uint256 delay;
+        if (customDelay[safe] > 0) {
+            delay = customDelay[safe];
+        } else {
+            delay = defaultDelay;
+        }
 
-        ISafe(safe).execTransactionFromModule(ecdsaPlugin, 0, data, 0);
+        recoveryRequests[safe].executeAfter = block.timestamp + delay;
+        recoveryRequests[safe].pendingNewOwner = newOwner;
+
+        emit RecoveryInitiated(safe, newOwner, block.timestamp + delay);
+    }
+
+    /**
+     * @notice recovers a safe ecdsa plugin using a zk email proof.
+     * @dev rotates the safe ecdsa plugin owner address to a new address.
+     * This function is designed so it can be called from any account and account type.
+     * @param safe the safe that manages the safe ecdsa plugin being recovered
+     * @param ecdsaPlugin safe ecsda plugin address that this function will be rotating the owner address for
+     */
+    function recoverPlugin(address safe, address ecdsaPlugin) external {
+        RecoveryRequest memory recoveryRequest = recoveryRequests[safe];
+
+        if (recoveryRequest.executeAfter == 0) {
+            revert RECOVERY_NOT_INITIATED();
+        }
+
+        if (block.timestamp > recoveryRequest.executeAfter) {
+            delete recoveryRequests[safe];
```

Comment:
Was torn on whether we should delete the mapping value after recovering the plugin. Curious if anyone had opinions on whether this was worthwhile.

Reply:
The question is, would someone want to use the same recovery configuration across multiple recoveries? I think the answer is yes. You should be able to recover using a specific email as many times as you want until such time as that recovery configuration is updated or removed. This is very similar to how existing password resets work: You can reset as many times as you want with an email until you update that email.

This also brings up another interesting question which is: Should we have a delay on adding, updating, or removing the recovery config as well. There could be cases where depending on the access restrictions/pattern we have on modifying the recovery config an attacker could update the config as the first txn of the attack, rendering the recovery useless. This also might make it worthwhile to think about locking/freezing the safe while the recovery process is ongoing.

I think all of the above is out of scope for this work, but is worth considering for future improvements to harden this recovery process. I would leave this as-is.

Reply:
Yeah good point. Added some notes on recovery hardening to a new issue that can be broken down https://github.com/getwax/wax/issues/184